service/worker/failoverManager/workflow.go

```diff
@@ -0,0 +1,376 @@
+		StartToCloseTimeout:    10 * time.Second,
+	}
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	} else {
+		return b
+	}
+}
+
+func validateParams(params *FailoverParams) error {
+	if params == nil {
+		return errors.New(errMsgParamsIsNil)
+	}
+	if params.BatchFailoverSize <= 0 {
+		params.BatchFailoverSize = defaultBatchFailoverSize
+	}
+	if params.BatchFailoverWaitTimeInSeconds <= 0 {
+		params.BatchFailoverWaitTimeInSeconds = defaultBatchFailoverWaitTimeInSeconds
+	}
+	return validateTargetAndSourceCluster(params.TargetCluster, params.SourceCluster)
+}
+
+// GetDomainsActivity activity def
+func GetDomainsActivity(ctx context.Context, params *GetDomainsActivityParams) ([]string, error) {
+	err := validateGetDomainsActivityParams(params)
+	if err != nil {
+		return nil, err
+	}
+	domains, err := getAllDomains(ctx, params.Domains)
+	if err != nil {
+		return nil, err
+	}
+	var res []string
+	for _, domain := range domains {
+		if shouldFailover(domain, params.SourceCluster) {
+			domainName := domain.GetDomainInfo().GetName()
+			res = append(res, domainName)
+		}
+	}
+	return res, nil
+}
+
+func validateGetDomainsActivityParams(params *GetDomainsActivityParams) error {
+	if params == nil {
+		return errors.New(errMsgParamsIsNil)
+	}
+	return validateTargetAndSourceCluster(params.TargetCluster, params.SourceCluster)
+}
+
+func validateTargetAndSourceCluster(targetCluster, sourceCluster string) error {
+	if len(targetCluster) == 0 {
+		return errors.New(errMsgTargetClusterIsEmpty)
+	}
+	if len(sourceCluster) == 0 {
+		return errors.New(errMsgSourceClusterIsEmpty)
+	}
+	if sourceCluster == targetCluster {
+		return errors.New(errMsgTargetClusterIsSameAsSource)
+	}
+	return nil
+}
+
+func shouldFailover(domain *shared.DescribeDomainResponse, sourceCluster string) bool {
+	currentActiveCluster := domain.ReplicationConfiguration.GetActiveClusterName()
+	isDomainTarget := currentActiveCluster == sourceCluster
+	return isDomainTarget && isDomainFailoverManagedByCadence(domain)
+}
+
+func isDomainFailoverManagedByCadence(domain *shared.DescribeDomainResponse) bool {
+	domainData := domain.DomainInfo.GetData()
+	return strings.ToLower(strings.TrimSpace(domainData[common.DomainDataKeyForManagedFailover])) == "true"
+}
+
+func getClient(ctx context.Context) frontend.Client {
+	manager := ctx.Value(failoverManagerContextKey).(*FailoverManager)
+	feClient := manager.clientBean.GetFrontendClient()
+	return feClient
+}
+
+func getAllDomains(ctx context.Context, targetDomains []string) ([]*shared.DescribeDomainResponse, error) {
+	feClient := getClient(ctx)
+	var res []*shared.DescribeDomainResponse
+
+	isTargetDomainsProvided := len(targetDomains) == 0
+	targetDomainsSet := make(map[string]struct{})
+	if isTargetDomainsProvided {
+		for _, domain := range targetDomains {
+			targetDomainsSet[domain] = struct{}{}
+		}
+	}
+
+	pagesize := int32(200)
+	var token []byte
+	for more := true; more; more = len(token) > 0 {
```

Comment:
nit: we do have a pagination iterator in common package

Reply:
I feel this is simple enough so will keep it, thanks for the info though.